dotnet/src/SemanticKernel.Abstractions/Kernel.cs

```diff
@@ -318,6 +327,82 @@ public CultureInfo Culture
     }
     #endregion
 
+    #region Internal Filtering
```

Comment:
It's good question whether it should be possible to add/remove events after Kernel initialization. I think the goal is to configure Kernel (e.g. in `Startup.cs`), add necessary filters and receive a call on specific event. But if we want to add/remove filters in the middle of runtime, then it's definitely not supported at the moment. 

If this is a requirement, we can add a possibility to add filter with something like this `kernel.FunctionFilters.Add(filter)` and `kernel.PromptFilters.Add(filter)`. But I'm wondering if it's useful at all. In theory, filters can be complex, include logger factory and other services to track function execution, and usually you want to configure these things on application startup only once. But maybe I'm missing specific use-case, please let me know.

Reply:
> It's good question whether it should be possible to add/remove events after Kernel initialization.

My point is that you _can_ currently add/remove events after Kernel initialization... in fact that's the only way you can do it. But you can't for filters.

Reply:
In general, it would be interesting to learn which way of event initialization is better in terms of usability - before Kernel initialization or after. For example, if we have ASP.NET app, controller and action, where exactly we attach an event, in controller constructor or action? Should we detach it at the end of action execution? Those usage patterns will allow us to understand with which approach we should stay in long-term.

Reply:
> Those usage patterns will allow us to understand with which approach we should stay in long-term.

Yes, but how will we determine that from this experiment? You can't register an event handler pre construction and you can't register a filter post construction. So if someone is using the events, are they doing so because it's more natural for their use case, or because they needed to register it post construction. And if someone is using the filters, are they doing so because it's more natural for their use case, or because the needed to register it pre construction?

Reply:
If I understand @stephentoub's points, it would be that we should _completely_ remove events (or deprecate them) when we add filters to make sure folks yell at us now in case they want to add events after kernel construction. Is that correct?

Reply:
Yes, _if_ that's how we intend to actually ship. Basically let's make it now how we believe it should be. If we believe long term we want both events and filters, then leave it as this PR has it. If we believe long term we only want filters, then I'd recommend removing the events now. Otherwise, it's harder for us to learn whether what we're shipping is sufficient.

Reply:
@stephentoub As part of this PR, I'm going to add a possibility to add filters to Kernel post construction, in order to cover case from events. Instead of removing events now, can we mark them as Obsolete to do a smooth rollout and avoid breaking changes?

Reply:
@stephentoub In this commit https://github.com/microsoft/semantic-kernel/pull/4437/commits/8ca9896fc64d1543e088506fe8b50621d777ef73 I've added collections for function and prompt filters and the ability to mutate filter lists after kernel registration. You can also insert filter in the middle of filter list and the order of filter execution will be updated as well. I've added related unit tests for new cases. Please let me know what you think about it.